arangod/IResearch/IResearchAqlAnalyzer.cpp

```diff
@@ -0,0 +1,516 @@
+  }
+
+  /// @brief create a transaction::Context
+  virtual std::shared_ptr<arangodb::transaction::Context> newTrxContext() const override {
+    return std::shared_ptr<arangodb::transaction::Context>(
+        std::shared_ptr<arangodb::transaction::Context>(), &_transactionContext);
+  }
+
+  virtual arangodb::transaction::Methods& trxForOptimization() override {
+    return *_trx;
+  }
+
+  virtual bool killed() const override { return false; }
+
+  /// @brief whether or not a query is a modification query
+  virtual bool isModificationQuery() const noexcept override { return false; }
+
+  virtual bool isAsyncQuery() const noexcept override { return false; }
+
+  virtual void enterV8Context() override {
+    TRI_ASSERT(FALSE);
+    THROW_ARANGO_EXCEPTION_MESSAGE(
+        TRI_ERROR_NOT_IMPLEMENTED,
+        "CalculationQueryContext entering V8 context is not implemented");
+  }
+
+ private:
+  arangodb::aql::QueryOptions _queryOptions;
+  arangodb::CollectionNameResolver _resolver;
+  mutable CalculationTransactionContext _transactionContext;
+  std::unique_ptr<arangodb::transaction::Methods> _trx;
+};
+
+frozen::set<irs::string_ref, 4> forbiddenFunctions{"TOKENS", "NGRAM_MATCH",
+                                                   "PHRASE", "ANALYZER"};
+
+arangodb::Result validateQuery(std::string const& queryStringRaw, TRI_vocbase_t& vocbase) {
+  try {
+    CalculationQueryContext queryContext(vocbase);
+    auto queryString = arangodb::aql::QueryString(queryStringRaw);
+    auto ast = queryContext.ast();
+    TRI_ASSERT(ast);
+    Parser parser(queryContext, *ast, queryString);
+    parser.parse();
+    ast->validateAndOptimize(queryContext.trxForOptimization());
+    AstNode* astRoot = const_cast<AstNode*>(ast->root());
+    TRI_ASSERT(astRoot);
+    // Forbid all V8 related stuff as it is not available on DBServers where analyzers run.
+    if (ast->willUseV8()) {
+      return {TRI_ERROR_BAD_PARAMETER,
+              "V8 usage is forbidden for calculation analyzer"};
+    }
+
+    // no modification (as data access is forbidden) but to give more clear error message
+    if (ast->containsModificationNode()) {
+      return {TRI_ERROR_BAD_PARAMETER,
+              "DML is forbidden for calculation analyzer"};
+    }
+
+    // no traversal (also data access is forbidden) but to give mo clear error message
+    if (ast->containsTraversal()) {
+      return {TRI_ERROR_BAD_PARAMETER,
+              "Traversal usage is forbidden for calculation analyzer"};
+    }
+
+    std::string errorMessage;
+    // Forbid to use functions that reference analyzers -> problems on recovery as analyzers are not available for querying.
+    // Forbid all non-Dbserver runable functions as it is not available on DBServers where analyzers run.
+    arangodb::aql::Ast::traverseReadOnly(
+        ast->root(), [&errorMessage](arangodb::aql::AstNode const* node) -> bool {
+          switch (node->type) {
+            case arangodb::aql::NODE_TYPE_FCALL: {
```

Comment:
It seems we are currently allowing subqueries in "aql" analyzers, which I don't think we should do.

Reply:
Why not after all ? Data access is forbidden but why to forbid calculating some filtered average for example ?

Reply:
Yes, for a filtered average, you will need either subqueries or `[* FILTER ...]`.
But this means you will also get the whole complexity of ExecutionPlans and ExecutionNodes, which you probably don't need by limiting it to `Expression`s. Anyway, if this is a requirement, so be it.

Reply:
i am also a bit resistant against allowing very complex things here.

Subqueries, For Loops, Sort, Collect I understand each of them is desirable to have, but will also create the possibility that we and up with very expensive analyzers, which may take up a lot of resources.

I am not finished with everything yet, and probably have not seen it, but I can imagine that it could get hard to stop / abort / identify / remove an analyzer that explodes in complexity and blocks a thread.